### PR TITLE
Bring back the PR `Refactor doctests + add CI` to `main`

### PR DESCRIPTION
**Pull request by ydshieh**
Reverts huggingface/transformers#23245

So we can keep the PR #22987 regarding the new doctest way, but without exposing `doctest_utils` to `src/transformers`.

@sgugger Let me know if you prefer to move this `doctest_utils.py` to `tests` folder.

### CODE REVIEW

**Review by sgugger — COMMENTED** [review]
Can you put the new testing utils (from `doctest_utils`) in `testing_utils`, so it all goes in the same place?

**Comment by ydshieh** [comment]
> Can you put the new testing utils (from `doctest_utils`) in `testing_utils`, so it all goes in the same place?

I am not against this suggestion, but I have some question about how to deal with the import of `pytest`.
In the current `testing_utils.py`, there is something like

```python
        try:
            import pytest  # We don't need a hard dependency on pytest in the main library
        except ImportError:
            return test_case
```

Do you suggest to put the import (in current `doctest_utils.py`) inside the classes/methods that need them?

```python
from _pytest.doctest import (
    Module,
    _get_checker,
    _get_continue_on_failure,
    _get_runner,
    _is_mocked,
    _patch_unwrap_mock_aware,
    get_optionflags,
    import_path,
)
from _pytest.outcomes import skip
from pytest import DoctestItem
```

**Comment by ydshieh** [comment]
> Can you put the new testing utils (from `doctest_utils`) in `testing_utils`, so it all goes in the same place?

In this case, am I allowed to put the import of `pytest` and `_pytest` on the top level of `testing_utils`? I am asking because I see in that file there is

```python
        try:
            import pytest  # We don't need a hard dependency on pytest in the main library
        except ImportError:
            return test_case
```

**Comment by sgugger** [comment]
There is no direct import into `testing_utils.py` so this should be fine to remove the try except (we will have until next release to make sure it doesn't create a new core dep of Transformers :sweat_smile: )

File: .circleci/config.yml [hunk 1]
```diff
@@ -166,6 +166,7 @@ jobs:
                       - v0.6-repository_consistency
             - run: pip install --upgrade pip
             - run: pip install .[all,quality]
+            - run: pip install pytest
```
Comment (sgugger) [inline .circleci/config.yml:169]: This shouldn't be necessary now, no?

**Review by sgugger — APPROVED** [review]
Thanks a lot!

**Comment by ydshieh** [comment]
This PR breaks
```bash
RUN_SLOW=1 python3 -m pytest -v tests/deepspeed/test_deepspeed.py::TrainerIntegrationDeepSpeed::test_can_resume_training_normal_zero2_fp16
```
(and some few others) with the error
```bash
PyTorch: setting up devices
PyTorch: setting up devices
PyTorch: setting up devices
....
....
(infinite)
FAILED tests/deepspeed/test_deepspeed.py::TrainerIntegrationDeepSpeed::test_can_resume_training_normal_zero2_fp16 - RecursionError: maximum recursion depth exceeded while calling a Python object
```
Need to check what's wrong.

**Comment by ArthurZucker** [comment]
Thanks for taking care of this! Think the filtered list could be obtain in a cleaner way with some bash commands but otherwise great 👍🏻

**Comment by ydshieh** [comment]
@ArthurZucker It was finally going to a `tests_fetcher.py`

https://github.com/huggingface/transformers/pull/23277/files

The bash command was just getting too complex ...

**Pull request merged**